scalanet/src/io/iohk/scalanet/codec/KMessageCodec.scala

```diff
@@ -0,0 +1,101 @@
+package io.iohk.scalanet.codec
+
+import java.nio.ByteBuffer
+import java.util.UUID
+
+import io.iohk.scalanet.peergroup.kademlia.{KMessage, KRouter}
+import scodec.Attempt.{Failure, Successful}
+import scodec.{Attempt, Codec, DecodeResult, Err, SizeBound}
+import scodec.bits.BitVector
+import scodec.codecs.implicits._
+
+class KMessageCodec[A](codec: Codec[KRouter.NodeRecord[A]])(implicit longCodec:Codec[Long],uuidCodec:Codec[UUID],byteCodec:Codec[Byte],bitVectorCodec:Codec[BitVector]) extends Codec[KMessage[A]]{
```

Comment:
Scodec uses shapeless for automatic deriviation of coproduct codecs. Take a look on object 
`io.iohk.scalanet.codec.DefaultCodecs.KademliaMessages`, it is enough to import this object into scope to derive codecs for `KMessage[A]` if there is codec for type `A`.